components/calendar/src/error.rs

```diff
@@ -59,13 +63,109 @@ pub enum DateError {
     /// ```
     #[displaydoc("Unknown month code {0:?}")]
     UnknownMonthCode(MonthCode),
+}
+
+impl core::error::Error for DateError {}
+
+/// Error type for date creation via [`Date::try_from_fields`].
+///
+/// [`Date::try_from_fields`]: crate::Date::try_from_fields
+#[derive(Debug, Copy, Clone, PartialEq, Display)]
+#[non_exhaustive]
+pub enum DateFromFieldsError {
+    /// A field is out of range for its domain.
+    #[displaydoc("{0}")]
+    Range(RangeError),
+    /// The era code is invalid for the calendar.
+    #[displaydoc("Unknown era or invalid syntax")]
+    UnknownEra,
+    /// The month code syntax is invalid.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::error::DateFromFieldsError;
+    /// use icu_calendar::Iso;
+    /// use icu_calendar::types::MonthCode;
+    /// use icu_calendar::types::DateFields;
+    /// use tinystr::tinystr;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.extended_year = Some(2000);
+    /// fields.month_code = Some(MonthCode(tinystr!(4, "????")));
+    /// fields.day = Some(1);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields, Default::default(), Iso
+    /// )
+    /// .expect_err("month code is invalid");
+    ///
+    /// assert!(matches!(err, DateFromFieldsError::InvalidMonthCode));
+    /// ```
+    #[displaydoc("Invalid month code syntax")]
+    InvalidMonthCode,
+    /// The specified month code does not exist in this calendar.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_calendar::Date;
+    /// use icu_calendar::error::DateFromFieldsError;
+    /// use icu_calendar::cal::Hebrew;
+    /// use icu_calendar::types::DateFields;
+    /// use icu_calendar::types::MonthCode;
+    /// use tinystr::tinystr;
+    ///
+    /// let mut fields = DateFields::default();
+    /// fields.extended_year = Some(5783);
+    /// fields.month_code = Some(MonthCode::new_normal(13).unwrap());
+    /// fields.day = Some(1);
+    ///
+    /// let err = Date::try_from_fields(
+    ///     fields,
+    ///     Default::default(),
+    ///     Hebrew
+    /// )
+    /// .expect_err("no month M13 in Hebrew");
+    /// assert!(matches!(err, DateFromFieldsError::UnknownMonthCodeForCalendar(_)));
+    /// ```
+    #[displaydoc("Month code does not exist in this calendar: {0:?}")]
+    UnknownMonthCodeForCalendar(MonthCode),
```

Comment:
thought: It would also be an okay API to not expose the month code here, the user can get to it anyway. But it's probably helpful to have? Maybe you have an argument to provide in either direction.

The main argument I see is you won't need to chase down an invalid month code and will be able to see what the month code was in your logs.